### DISCUSSION TITLE
Apps as a Service

### DISCUSSION THREAD

**Discussion by madsolar8582** [MobileNativeFoundation/discussions#70] [General]
Hello, my name is Madison and I am the engineer responsible for my company's (Healthcare I.T.) mobility platform. We provide the base frameworks and application skeletons for both the iOS and Android platforms. 

Today, we have ~14 apps for each platform and a medium-sized library of reusable components. The company has been moving towards Continuous Delivery and that has met some challenges in the mobility space. Therefore, my team has been given the directive to improve the operational efficiency of the teams that own the applications and components. Primarily, this directive has led us down the route of updating our Platform as a Service to be Applications as a Service.

### Application Generators

The apps today have been customized to fit business needs, so they don't particularly resemble the skeleton from whence they were generated. One of the ideas we are floating is to extend our current skeleton generation process to be able to generate applications (alongside their test targets, extensions, resource bundles, etc.) on demand. On Android, this seems more feasible as the manifests are all XML-based and gradle has a well defined DSL. However, on iOS it is looking quite complicated and error prone due to proprietary file formats as well as brittle since Apple tends to change things quite frequently. Has anyone gone down this path? If so, how did you accomplish it?

### Mono Projects

Presuming full on regeneration of all artifacts is not feasible (or prohibitively expensive and complex), another idea we’ve been thinking about is having all apps live in a single project (kinda like a mono repo). However, we're concerned about this scaling. For the iOS space, how does Xcode handle a large number of targets (i.e. app, extensions, test, resource bundles, etc.) and does CocoaPods perform well? Similarly on Android, does Android Studio handle this well? Does gradle keep its memory usage reasonable and can it finish tasks in a reasonable amount of time taking into account desugaring? If anyone has done this, what are your experiences?

### Task Automation

Lastly, we were thinking that if any of our grand ideas above are too complex, expensive, or do not solve the entire problem, we could create bots that would take care of maintenance tasks (e.g. dependency updates). For those who have automated bots, did you build your own or are you using something like Dependabot? What difficulties or challenges have you encountered?

**Comment by Alex009** [comment]
hello, @madsolar8582 ! i am Aleksey Mikhailov from IceRock. I can suggest to you application generator. to be more precise, it is not only a generator of applications. It's generator for any directory structures with handlebars template engine. 
https://github.com/icerockdev/shaper - tool, with cli for generation
and here https://github.com/icerockdev/shaper-templates contained template for our moko libraries projects. 
i think that this tool can generate any project/module/small piece of code, with some template-defined arguments. check it and maybe it will help you!

**Reply by madsolar8582** [reply]
Thank you for the recommendation Aleksey, we'll definitely take a look.

**Comment by ob** [comment]
We at LinkedIn have automated this in pretty much the same way you mention. We have an application generator that creates an app that has basic functionality wired in (including tests) and can be run through our CI pipeline end to end. We also have a dependency updater that we wrote to keep libraries and external frameworks up-to-date. We don't have a monorepo, each app is a git repository of its own.

If you went the monorepo route, especially with a large number of apps, it would probably be better to build the apps with Bazel instead of Xcode. I would not recommend the monorepo though... I feel the scaling problems are harder in a monorepo scenario than in a multi-repo one.

**Reply by madsolar8582** [reply]
Thanks for the insights Oscar. If you don't mind me asking, what language is your updater written in? We do a lot of Rails work, so we were thinking Ruby since GitHub has OctoKit available as a gem.

**Reply by ob** [reply]
In our case, since we are reusing the same infra for many other products, it's a mixture of Python and Java. If it was mobile specific Ruby seems like a sensible choice.

**Reply by omiz** [reply]
Hi Oscar, I'm interested in the dependency updater you mentioned Is it a smart to update the code when needed? I would appreciate it if you can share your experience.

**Reply by ob** [reply]
It's not super-smart... what it does is that it bumps the dependencies to the latest (respecting semver) and runs a full build & test with the bumped dependencies. If there are incompatibilities the idea is that the build & test will break and we'll notice and inform the product owners. Otherwise the dependencies can be bumped and at least build & test works :-)

**Reply by donnie-jp** [reply]
Interesting!

Do the generated apps follow/include any particular app architecture?
Or is it more about having a skeleton that includes the standard dependencies?
Does it include UI code?

**Comment by donnie-jp** [comment]
Can anyone share any good blog posts etc. that go into more details about this concept?
